src/bytes.rs

```diff
@@ -0,0 +1,749 @@
+/// the highest-order bit of the InlineVec's `tagged_len` would be set. This is
+/// true because EcoVec is repr(C) and it's second field `len` is bounded by
+/// `isize::MAX`. On 32-bit, it's no problem and for 64-bit big endian, we
+/// have an increased limit to prevent the overlap.
+#[repr(C)]
+union Repr {
+    inline: InlineVec,
+    spilled: ManuallyDrop<EcoVec<u8>>,
+}
+
+/// This is never stored in memory, it's just an abstraction for safe access.
+#[derive(Debug)]
+enum Variant<'a> {
+    Inline(&'a InlineVec),
+    Spilled(&'a EcoVec<u8>),
+}
+
+/// This is never stored in memory, it's just an abstraction for safe access.
+#[derive(Debug)]
+enum VariantMut<'a> {
+    Inline(&'a mut InlineVec),
+    Spilled(&'a mut EcoVec<u8>),
+}
+
+/// The maximum amount of inline storage. Typically, this is 15 bytes.
+///
+/// However, in the rare exotic system, we still want things to be safe.
+/// Therefore, the following special cases:
+/// - For big endian, we increase the limit such that the tagged length of the
+///   inline variants doesn't overlap with the EcoVec. For little endian, its
+///   fine since the highest order bit is never set for a valid EcoVec.
+/// - In case somehow EcoVec is very big (128-bit pointers woah), increase the
+///   limit too.
+pub(crate) const LIMIT: usize = {
+    let mut limit = 15;
+    if limit < mem::size_of::<EcoVec<u8>>() - 1 {
+        limit = mem::size_of::<EcoVec<u8>>() - 1;
+    }
+    if cfg!(target_endian = "big") {
+        limit += mem::size_of::<usize>();
+    }
+    limit
+};
+
+/// This bit is used to check whether we are inline or not. On 64-bit little
+/// endian, it coincides with the highest-order bit of an EcoVec's length, which
+/// can't be set because the EcoVec's length never exceeds `isize::MAX`.
+const LEN_TAG: u8 = 0b1000_0000;
+
+/// This is used to mask off the tag to get the inline variant's length.
+const LEN_MASK: u8 = 0b0111_1111;
+
+impl EcoBytes {
+    /// Maximum number of bytes for an inline `EcoBytes` before spilling on
+    /// the heap.
+    ///
+    /// The exact value for this is architecture dependent.
+    ///
+    /// # Note
+    /// This value is semver exempt and can be changed with any update.
+    pub const INLINE_LIMIT: usize = LIMIT;
+
+    /// Create a new, empty vector.
+    #[inline]
+    pub const fn new() -> Self {
+        Self::from_inline(InlineVec::new())
+    }
+
+    /// Create a new, inline vector.
+    ///
+    /// Panics if the vector's length exceeds the capacity of the inline
+    /// storage.
+    #[inline]
+    pub const fn inline(bytes: &[u8]) -> Self {
+        let Ok(inline) = InlineVec::from_slice(bytes) else {
+            exceeded_inline_capacity();
+        };
+        EcoBytes::from_inline(inline)
+    }
+
+    /// Create a new, inline vector.
+    ///
+    /// Returns `Err(())` if the vector's length exceeds the capacity of
+    /// the inline storage.
+    #[inline]
+    pub const fn try_inline(bytes: &[u8]) -> Result<Self, ()> {
```

Comment:
I thinking returning `Option` would be preferrable over having `()` as an error.